source/beamline/AMMotorGroup.cpp

```diff
@@ -312,17 +805,59 @@ AMMotorGroup::AMMotorGroup(QObject *parent)
 
 }
 
-AMMotorGroupObject *AMMotorGroup::motorGroupObject(const QString &name) const
+QList<QString> AMMotorGroup::names() const
+{
+	return groupObjects_.keys();
+}
+
+QList<AMMotorGroupObject *> AMMotorGroup::motorGroupObjects() const
+{
+	return groupObjects_.values();
+}
+
+int AMMotorGroup::size() const
+{
+	return groupObjects_.size();
+}
+
+AMMotorGroupObject * AMMotorGroup::motorGroupObject(const QString &name) const
 {
-	return infoMap_.value(name, 0);
+	return groupObjects_.value(name, 0);
 }
 
 QString AMMotorGroup::name(AMMotorGroupObject *motorObject) const
 {
-	return infoMap_.key(motorObject, QString());
+	return groupObjects_.key(motorObject, QString());
 }
 
-void AMMotorGroup::addMotorGroupObject(const QString &name, AMMotorGroupObject *object)
+void AMMotorGroup::addMotorGroupObject(AMMotorGroupObject *object)
 {
-	infoMap_.insert(name, object);
+	if(object == 0) {
+		return;
+	}
+
+	AMMotorGroupObject* currentObject = motorGroupObject(object->name());
+
+	if(currentObject == object) {
```

Comment:
I think if you want to do something like this, you should let the caller deal with the memory.  I would change this to `bool` function and if it returns false, the caller will know that they need to deal with the object.